Pre-calculate all non-active buckets in Epsilon Greedy Pool

suhailpatel opened the pull request:
This PR makes some substantial performance improvements to the Epsilon Greedy Hostpool.

Before I explain the performance improvements, I need to explain a little about how the scoring calculation works per host. Let's take an example of a host with 5 stored buckets. The pool keeps track of two core values per bucket, a count and a cumulative duration total (in nanoseconds). 

The goal is to keep track of performance over time but also weigh real time performance higher when scoring.

The list of buckets is a fixed size ring buffer, here we use 5 (but in practice go-hostpool uses 120 buckets). 
- Each bucket has an proportionally reduced weighting depending on how old it is (so older buckets have a reduced weighting, more recent buckets have an increased weighting). 
- The current bucket has a 100% weighting to the score (as it is the most recent values)

![image](https://user-images.githubusercontent.com/76773/109498846-f4bcf880-7a8b-11eb-80ba-78ee5be81d45.png)

So in the example above, the score for this particular host will be calculated as (starting from the oldest bucket)
```
      6 ×  540 × (1 / 5.0)
  +  15 × 1495 × (2 / 5.0)
  +  11 ×  900 × (3 / 5.0)
  +  10 × 1500 × (4 / 5.0)
  +  12 × 1140 × (5 / 5.0)
```

When decay happens, the current bucket and oldest bucket offsets move one index forward, resetting the current bucket to zero.

After some profiling, I found this calculation to be taking a substantial part of CPU time and was being done _every time we requested a host_. I found that this entire calculation could be optimised substantially to pre-calculate most of it. Only the average of the current bucket is changing in this calculation. We can precompute all the other bucket totals and just add the current bucket average on demand

![image](https://user-images.githubusercontent.com/76773/109500075-a7418b00-7a8d-11eb-88f6-ffdb5d7af1ce.png)

This means on each decay, we pre-compute the following sum one time on decay and essentially 'cache' it
```
      6 ×  540 × (1 / 5.0)
  +  15 × 1495 × (2 / 5.0)
  +  11 ×  900 × (3 / 5.0)
  +  10 × 1500 × (4 / 5.0)
```

Then when we request the current score of the host, we take the current bucket average (which is weighted at 100% anyway) and add it on, thus saving four other calculations in this example (and 119 per host in real life).

Before these changes, here's how the performance stacked up
```
pkg: github.com/bitly/go-hostpool
BenchmarkEpsilonGreedy
BenchmarkEpsilonGreedy-8            	  449230	      2820 ns/op	     127 B/op	       3 allocs/op
BenchmarkEpsilonGreedyManyHosts
BenchmarkEpsilonGreedyManyHosts/Hosts10
BenchmarkEpsilonGreedyManyHosts/Hosts10-8         	  105246	     12184 ns/op	     349 B/op	       6 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25
BenchmarkEpsilonGreedyManyHosts/Hosts25-8         	   42470	     29409 ns/op	     601 B/op	       7 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50
BenchmarkEpsilonGreedyManyHosts/Hosts50-8         	   21153	     59977 ns/op	    1070 B/op	       8 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100
BenchmarkEpsilonGreedyManyHosts/Hosts100-8        	   10000	    114046 ns/op	    1663 B/op	       9 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250
BenchmarkEpsilonGreedyManyHosts/Hosts250-8        	    5290	    268156 ns/op	    1047 B/op	       8 allocs/op
```

This meant if you have 100 hosts, you'll be burning 114046 nanoseconds of CPU time (under a mutual lock also) fetching a host. That's a very long amount of time!

With these changes, things get substantially better!
```
pkg: github.com/bitly/go-hostpool
BenchmarkEpsilonGreedy
BenchmarkEpsilonGreedy-8            	 2128981	       559 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts
BenchmarkEpsilonGreedyManyHosts/Hosts10
BenchmarkEpsilonGreedyManyHosts/Hosts10-8         	 1425793	       768 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25
BenchmarkEpsilonGreedyManyHosts/Hosts25-8         	 1000000	      1115 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50
BenchmarkEpsilonGreedyManyHosts/Hosts50-8         	  739261	      1719 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100
BenchmarkEpsilonGreedyManyHosts/Hosts100-8        	  474729	      3141 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250
BenchmarkEpsilonGreedyManyHosts/Hosts250-8        	  246646	      6379 ns/op	     104 B/op	       2 allocs/op
```

That's a 30x to 50x improvement by being a bit more intelligent!

suhailpatel committed:
Initialize with a go.mod file

suhailpatel committed:
Update travis config to test on latest Go versions

suhailpatel committed:
Use subtests for working with many hosts

suhailpatel committed:
Move the heavy lifting of calculating the epsilon values to the decay function

Prior to this change, we were calculating the weighted values of each
bucket every time we asked for a host. That is incredibly wasteful
because we can pre-calculate averages of all buckets except the current
bucket each time we decay.

This results in a significant speed improvement

suhailpatel commented on epsilon_greedy.go:
I got rid of this `epsilonPercentage` value. It's a local value to this calculation so has no reason being part of that struct. It's unclean and also means in future it's as difficult to get rid of the mutual lock

suhailpatel commented on epsilon_greedy.go:
```diff
@@ -108,10 +108,7 @@ func (p *epsilonGreedyHostPool) epsilonGreedyDecay() {
 func (p *epsilonGreedyHostPool) performEpsilonGreedyDecay() {
 	p.Lock()
 	for _, h := range p.hostList {
-		h.epsilonIndex += 1
```
This has been moved into the epsilonDecay function

suhailpatel committed:
Measure with and without decay as a factor

kjgorman commented on host_entry.go:
```diff
@@ -42,25 +43,49 @@ func (h *hostEntry) willRetryHost(maxRetryInterval time.Duration) {
 func (h *hostEntry) getWeightedAverageResponseTime() float64 {
-	var value float64
-	var lastValue float64
+	currentBucketCount := h.epsilonCounts[h.epsilonIndex]
```
I added some test cases in `host_entry_test.go` over in [this branch](https://github.com/monzo/go-hostpool/compare/master...monzo:kjgorman/configurables?expand=1) where I was extracting some properties (like bucket counts) — they highlight that in a zero case of no hosts this line can panic

suhailpatel replied:
This panic will occur if the hostEntry hasn't been set up properly. if we look at where it is set up for use in epsilon greedy, [it is initialised with the zero values for all the buckets it expects](https://github.com/monzo/go-hostpool/blob/master/epsilon_greedy.go#L65-L69).

At this point, I feel like an internal invariant has been broken and it _should_ panic rather than paper over the crack by adding a condition within the calculation function (which would not panic but would mask another bug). Given the implementation is entirely scoped and private to within this project, it's fair to hold the invariant.

This is solved in the test by tweaking it to
```			
entry: &hostEntry{
	epsilonCounts: []int64{0},
	epsilonValues: []int64{0},
	epsilonIndex:  0,
},
```

suhailpatel commented:
I tweaked the benchmark a little because we're decaying after every 100 ticks which in the much faster version is quite often and that was skewing the results. The decay in reality occurs once on a timer as mentioned on the original post and not every time we request a host.

Here's the benchmark results with decay disabled and decay enabled.

Before
```
pkg: github.com/bitly/go-hostpool
BenchmarkEpsilonGreedyManyHosts
BenchmarkEpsilonGreedyManyHosts/Hosts10/NoDecay-8         	  109764	     11249 ns/op	     349 B/op	       6 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25/NoDecay-8         	   44950	     27166 ns/op	     602 B/op	       7 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50/NoDecay-8         	   23008	     51590 ns/op	    1084 B/op	       8 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100/NoDecay-8        	   10000	    102160 ns/op	    1757 B/op	       9 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250/NoDecay-8        	    5642	    240300 ns/op	    1428 B/op	       9 allocs/op

BenchmarkEpsilonGreedyManyHosts/Hosts10/WithDecay-8       	  102639	     11637 ns/op	     349 B/op	       6 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25/WithDecay-8       	   44194	     28124 ns/op	     601 B/op	       7 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50/WithDecay-8       	   21825	     57196 ns/op	    1095 B/op	       8 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100/WithDecay-8      	   10000	    111213 ns/op	    1648 B/op	       9 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250/WithDecay-8      	    5392	    247852 ns/op	    1220 B/op	       8 allocs/op
```

After
```
pkg: github.com/bitly/go-hostpool
BenchmarkEpsilonGreedyManyHosts
BenchmarkEpsilonGreedyManyHosts/Hosts10/NoDecay-8         	 2070523	       574 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25/NoDecay-8         	 1719786	       660 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50/NoDecay-8         	 1417652	       841 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100/NoDecay-8        	 1000000	      1280 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250/NoDecay-8        	  584746	      2701 ns/op	     104 B/op	       2 allocs/op

BenchmarkEpsilonGreedyManyHosts/Hosts10/WithDecay-8       	 1878009	       643 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts25/WithDecay-8       	 1238664	       923 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts50/WithDecay-8       	  924391	      1439 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts100/WithDecay-8      	  569004	      2708 ns/op	     104 B/op	       2 allocs/op
BenchmarkEpsilonGreedyManyHosts/Hosts250/WithDecay-8      	  282620	      5691 ns/op	     104 B/op	       2 allocs/op
```

Looking at the decay disabled numbers for comparison, that's an ~85x improvement for the hot path 🤯

kjgorman commented on epsilon_greedy.go:
is doing `1.0 / hostResponseTime` here effectively hard-coding what was previously in `CalcValueFromAvgResponseTime`? Do we need to remove the `EpsilonValueCalculator` if we are not using it?

suhailpatel replied:
It should use `CalcValueFromAvgResponseTime`, i've fixed it 👍

kjgorman reviewed (commented):
Looks good to me—small question on an empty case and whether ignoring the calculator is perfectly backwards compatible, but mostly makes sense.

Is the intuition that the active bucket is the only thing which needs to be checked on the 'hot' path, and we can shift the calculation of all other buckets into the background decay step?

suhailpatel committed:
Add some tests for our host entry that exercise our calculation

Tests and asserts against some pre-calculated maths that should be
deterministic.

Co-authored-by: Kieran Gorman <kieran@kjgorman.com>

suhailpatel commented:
I've fixed the use of the calculator, that was some testing on my part, we should use the calculator provided especially because that's part of the exposed interface

> Is the intuition that the active bucket is the only thing which needs to be checked on the 'hot' path, and we can shift the calculation of all other buckets into the background decay step?

Exactly, we keep a running total with everything else in the decay step, then we can just 'append' the calculation for the current bucket to come up with the overall score.

kjgorman reviewed (approved):
lgtm — excited to see some 📉 when we give it a go in hot service 🚀

suhailpatel committed:
Only need to call time once
